File: src/main/java/AoC2016_08.java
import static com.github.pareronia.aoc.IntegerSequence.Range.range;
import static com.github.pareronia.aoc.StringOps.splitLines;
import static com.github.pareronia.aoc.StringOps.splitOnce;
import static java.util.stream.Collectors.toSet;

import java.util.List;
import java.util.Set;

import com.github.pareronia.aoc.CharGrid;
import com.github.pareronia.aoc.Grid.Cell;
import com.github.pareronia.aoc.OCR;
import com.github.pareronia.aoc.StringOps.StringSplit;
import com.github.pareronia.aoc.itertools.IterTools;
import com.github.pareronia.aoc.StringUtils;
import com.github.pareronia.aoc.solution.SolutionBase;

public class AoC2016_08 extends SolutionBase<List<String>, Integer, String> {

    private static final char ON = '#';
	private static final char OFF = '.';
    
	private AoC2016_08(final boolean debug) {
		super(debug);
	}
	
	public static final AoC2016_08 create() {
		return new AoC2016_08(false);
	}

	public static final AoC2016_08 createDebug() {
		return new AoC2016_08(true);
	}
    
    @Override
    protected List<String> parseInput(final List<String> inputs) {
        return inputs;
    }
	
    private CharGrid solve(
            final List<String> inputs,
            final Integer rows,
            final Integer columns
    ) {
	    CharGrid grid = CharGrid.from(
	            range(rows).intStream()
	                .mapToObj(i -> StringUtils.repeat(OFF, columns))
	                .toList());
	    for (final String input : inputs) {
            if (input.startsWith("rect ")) {
                final StringSplit coords = splitOnce(
                        input.substring("rect ".length()), "x");
                final Set<Cell> cells = IterTools.product(
                        range(Integer.parseInt(coords.right())),
                        range(Integer.parseInt(coords.left()))).stream()
                    .map(lst -> Cell.at(lst.first(), lst.second()))
                    .collect(toSet());
                grid = grid.update(cells, ON);
            } else if (input.startsWith("rotate row ")) {
                final StringSplit coords = splitOnce(
                        input.substring("rotate row ".length()), " by ");
                final int row = Integer.parseInt(coords.left().substring("y=".length()));
                final int amount = Integer.parseInt(coords.right());
                grid = grid.rollRow(row, amount);
            } else {
                final StringSplit coords = splitOnce(
                        input.substring("rotate column ".length()), " by ");
                final int column = Integer.parseInt(coords.left().substring("x=".length()));
                final int amount = Integer.parseInt(coords.right());
                grid = grid.rollColumn(column, amount);
            }
            log("");
            log(grid);
        }
	    return grid;
	}

	@Override
	public Integer solvePart1(final List<String> inputs) {
	    return (int) solve(inputs, 6, 50).countAllEqualTo(ON);
	}
	
	@Override
	public String solvePart2(final List<String> inputs) {
	    return OCR.convert6(solve(inputs, 6, 50), ON, OFF);
	}

    @Override
    public void samples() {
        final AoC2016_08 test = AoC2016_08.createDebug();
        assert test.solve(test.parseInput(splitLines(TEST)), 3, 7)
                .countAllEqualTo(ON) == 6;
    }

    public static void main(final String[] args) throws Exception {
        AoC2016_08.create().run();
	}

	private static final String TEST = """
	        rect 3x2
	        rotate column x=1 by 1
	        rotate row y=0 by 4
	        rotate column x=1 by 1
	        """;
}

File: src/main/python/AoC2016_08.py
#! /usr/bin/env python3
#
# Advent of Code 2016 Day 8
#

import itertools
import sys
from typing import cast

from advent_of_code_ocr import convert_6
from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.grid import Cell
from aoc.grid import CharGrid

Input = InputData
Output1 = int
Output2 = str

ON = "#"
OFF = "."

TEST = """\
rect 3x2
rotate column x=1 by 1
rotate row y=0 by 4
rotate column x=1 by 1
"""


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return input_data

    def solve(self, input: Input, rows: int, columns: int) -> CharGrid:
        grid = CharGrid.from_strings([OFF * columns for _ in range(rows)])
        for line in input:
            if line.startswith("rect "):
                y, x = map(int, line[len("rect ") :].split("x"))  # noqa E203
                for r, c in itertools.product(range(x), range(y)):
                    grid.set_value(Cell(r, c), ON)
            elif line.startswith("rotate row "):
                offset = len("rotate row ")
                row, amount = line[offset:].split(" by ")
                grid.roll_row(int(row[2:]), int(amount))
            else:
                offset = len("rotate column ")
                column, amount = line[offset:].split(" by ")
                grid.roll_column(int(column[2:]), int(amount))
        return grid

    def solve_1(self, input: Input, rows: int, columns: int) -> Output1:
        grid = self.solve(input, rows, columns)
        return len(list(grid.get_all_equal_to(ON)))

    def part_1(self, input: Input) -> Output1:
        return self.solve_1(input, 6, 50)

    def part_2(self, input: Input) -> Output2:
        grid = self.solve(input, 6, 50)
        return cast(
            str,
            convert_6(
                "\n".join(grid.get_rows_as_strings()),
                fill_pixel=ON,
                empty_pixel=OFF,
            ),
        )

    def samples(self) -> None:
        assert self.solve_1(self.parse_input(TEST.splitlines()), 3, 7) == 6


solution = Solution(2016, 8)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()
